### ISSUE TITLE
Adding a vapi to 'sources' promote the depending library to a Vala target

### ISSUE THREAD

**Issue by arteymix** [mesonbuild/meson#1307]
Here's the typical use case where you create some library and bundle a vapi along, potentially from `vapigen`, and use it along a C target:

```python
foo_dep = declare_dependency(link_with: foo_lib, sources: foo_vapi)
bar = executable('bar', 'bar.c', dependency: foo_dep)
```

This is causing the following error:

```
Vala library has no Vala source files.
```

I'm either doing this wrong or having a `*.vapi` as source should not be a sufficient criteria to promote a target.

**Comment by nirbheek** [comment]
You're right, this is a bug. However, this is partly because the way we declare a Vala dependency is hacky. The correct fix is to make `declare_dependency` understand Vala dependencies. We will need this for Rust too.
